src/docfx/build/toc/parser/TableOfContentsParser.cs

```diff
@@ -11,23 +11,32 @@ namespace Microsoft.Docs.Build
 {
     internal static class TableOfContentsParser
     {
-        internal static List<TableOfContentsItem> Load(string tocContent, bool isYaml, Document filePath = default, Document rootPath = default, ResolveContent resolveContent = null, ResolveLink resolveLink = null)
+        internal static List<TableOfContentsItem> Load(string tocContent, bool isYaml, Document filePath, Document rootPath = default, ResolveContent resolveContent = null, ResolveLink resolveLink = null, HashSet<Document> parents = null)
         {
-            var models = isYaml ? LoadYamlTocModel(tocContent, filePath) : LoadMdTocModel(tocContent);
+            parents = parents ?? new HashSet<Document>();
+
+            // add to parent path
+            if (!parents.Add(filePath))
+            {
+                // todo: error handling
+                throw new ApplicationException($"Circle toc reference was detected, {string.Join("-->", parents)}-->{filePath}");
```

Comment:
`HashSet` is not ordered, this could be wrong. It should be a stack.

Reply:
ok!

---
In reply to: [185987467](https://github.com/dotnet/docfx/pull/2725#discussion_r185987467) [](ancestors = 185987467)